polar-core/src/vm.rs

```diff
@@ -2158,8 +2138,19 @@ impl PolarVirtualMachine {
             check_applicability.push(Goal::Cut {
                 choice_index: self.choices.len(),
             });
-            check_applicability.push(applicable);
-            self.choose(vec![check_applicability, vec![inapplicable]])?;
+            check_applicability.push(Goal::Backtrack);
+
+            // If a check fails, we will backtrack to the inapplicable branch.
+            self.choose(vec![
```

Comment:
Maybe replace this with a `choose_exactly_one` or something? Is that the intention? Or something like a choose_xor? Do applicable XOR inapplicable?

Reply:
I could see it as something like `push_conditional(conditions: Vec<Goal>, consequent: Goal, alternative: Goal)`, so in this case:

```rust
push_conditional(check_applicability, applicable, inapplicable)
```

Reply:
Looks like @dhatch had similar thoughts: https://github.com/osohq/oso/pull/420#pullrequestreview-488075393

Reply:
Looks _so_ much better with this